torch/fx/experimental/proxy_tensor.py

```diff
@@ -170,15 +187,34 @@ def wrapped(*args):
     return wrapped
 
 
-def make_fx(f, decomposition_table=None):
+class ProxyTorchDispatchMode(TorchDispatchMode):
+    def __init__(self, tracer):
+        self.tracer = tracer
+
+    def __torch_dispatch__(self, func_overload, types, args=(), kwargs=None):
```

Comment:
Hi, @samdow 
May i ask some questions about code here? I am trying to understand the logic here.
- What is `ProxyTorchDispatchMode` used for? Is it used to create the FX proxy node(only for torch aten operator) onto the FX graph?
- I wonder which op will call `__torch_dispatch__` here? I found no python API explicitly call this interface.
- For the runtime op on the FX graph, here `ProxyTorchDispatchMode` cannot handle it right?

Reply:
> Is it used to create the FX proxy node(only for torch aten operator) onto the FX graph through calling the func_overload here

Yes, it's used to capture FX proxy nodes.

> I wonder which function will call __torch_dispatch__ here

Yes, it's done in the C++ dispatcher (through the Python C++ dispatch key).

> For the runtime op on the FX graph, here ProxyTorchDispatchMode cannot handle and trace them onto the FX graph right?

Yes, they're "traced through" onto the graph. That is, they are not explicitly represented, but their effects are captured through the aten nodes that are captured.

Reply:
Hi, @Chillee 

Thank you so much for answers. It helps lot. I go on dive the code and have some confusion.

- I notice there are `FakeTensorMode` context set for the `dispatch_trace` function. I wonder if this context is only used for tracing the graph when torch.compile(.., **dynamic=True**)? Does the AOTAutograd tracing only run all aten ops with the FakeTensor or real tensor?
(code here: https://github.com/pytorch/pytorch/blob/29844adbe0ad7437ff00f9a3eee544e4e94a10ed/torch/fx/experimental/proxy_tensor.py#L860)

- I notice the `tracing_mode` has 3 types: real, fake and symbolic. I set the torch.compile(.., **dynamic=True**), while the `tracing_mode` in `make_fx` is still real mode, not symbolic mode. I wonder what these 3 tracing modes represent?
(code here: https://github.com/pytorch/pytorch/blob/29844adbe0ad7437ff00f9a3eee544e4e94a10ed/torch/fx/experimental/proxy_tensor.py#L763)

- I find the `FunctionalTensorMode` also inherits from the `TorchDispatchMode` and overrides the `__torch_dispatch__`. I am confused about why torch needs this mode. The code PR says it is used for functionalization and this mode works with a special tensor called `FunctionalTensor`.
(code here: https://github.com/pytorch/pytorch/blob/29f3d392bf230072e3bffae37b078e770cae1956/torch/_subclasses/functional_tensor.py#L220)

- I also find there are 3 Torch Dispatch Mode Keys named below. The first one seems to create the proxy node onto the graph. How the keys here affect the AOTAutograd tracing?
```
torch._C._TorchDispatchModeKey.PROXY
torch._C._TorchDispatchModeKey.FAKE
torch._C._TorchDispatchModeKey.FUNCTIONAL
```

- The runtime ops are "traced through" onto the graph because all context(inherit from the `TorchDispatchMode`) are working for aten ops dispatch through `__torch_dispatch__`. Could there be an additional mode for tracing runtime ops onto the graph because some runtime ops cannot affect the aten nodes, they affect the execution behavior, for example, `torch.cuda.set_stream(...)`. It maybe hard to do because runtime op has no dispatch behavior and no tensors invovled. Why i want to do this because i found a issue here: https://github.com/pytorch/pytorch/issues/109848
(code here: https://github.com/pytorch/pytorch/blob/29844adbe0ad7437ff00f9a3eee544e4e94a10ed/torch/fx/experimental/proxy_tensor.py#L860)

Thank you. :)

Reply:
> Does the AOTAutograd tracing only run all aten ops with the FakeTensor or the real tensor?

>  I wonder what these 3 tracing modes represent?

The name on `make_fx` is a bit misleading, and is a bit of a hold-over from long ago when we were using `make_fx` directly without Dynamo. Nowadays, under `torch.compile`, the tensors that are passed into `make_fx` are faketensors, and so the trace always occur under a faketensormode. 

Concretely, those modes do additional wrapping on the tensors to convert them into faketensors. But under `torch.compile`, the input tensors are already fakified, so we just use `tracing_mode="real"`.

> I find the FunctionalTensorMode also inherits from the TorchDispatchMode and overrides the __torch_dispatch__. I am confused about why torch needs this mode.

If you're asking what functionalization does in general, it converts torch ops from mutating versions to non-mutating versions. So the graph we trace out after functionalization has no mutation in it. If you're asking why we need a python version of functionalization instead of just the C++ version that's a bit more nuanced of a question.

> How the keys here affect the AOTAutograd tracing?

First one is for tracing, second one is for metadata propagation, third one is for functionalization.

> Could there be an additional mode for tracing runtime ops onto the graph because some runtime ops cannot affect the aten nodes, but they affect the execution behavior, for example, torch.cuda.set_stream(...)?

Currently we do not have an easy way of tracing stream operators into the graph. There's been some discussions about it, but I'm not totally sure about the current status. I think the issue you linked talks about the current state.

Reply:
Hi, @Chillee 

Thank you for helpful answers. It solves lots of my confusion. 👍 

> If you're asking what functionalization does in general, it converts torch ops from mutating versions to non-mutating versions. So the graph we trace out after functionalization has no mutation in it.
- Here how can i understand the `mutation` above? The traced-out graph will have no mutation in it. Does it mean there are no aliasing tensor(placeholder) FX nodes in the graph? Why the torch ops need to be converted from mutating versions to non-mutating versions during AOTAutograd tracing?


- I have a quite simple case for dive `torch.compile` like below.
```
import torch

@torch.compile()
def test(x):
    x = x + 1.0
    x = x * 2.0
    x = x / 3.0
    return x

input = torch.randn(3, 3, device='cuda')
output = test(input)
```
I get the joint graph after the `dispatch_trace`. All 3 aten ops in graph are associated with the ops in the function `test`. It is expected.

```
def forward(self, arg0_1):
    add = torch.ops.aten.add.Tensor(arg0_1, 1.0);  arg0_1 = None
    mul = torch.ops.aten.mul.Tensor(add, 2.0);  add = None
    div = torch.ops.aten.div.Tensor(mul, 3.0);  mul = None
    return (div,)
```
While the AOTAutograd tracing seems not stop, the `dispatch_trace` is called again and return a graph like below.
```
def forward(self, primals, tangents):
    primals_1, primals_2, tangents_1, = fx_pytree.tree_flatten_spec([primals, tangents], self._in_spec)
    mm = torch.ops.aten.mm.default(primals_1, primals_2)
    is_same_size = torch.ops.aten.is_same_size.default(mm, tangents_1)
    permute = torch.ops.aten.permute.default(primals_1, [1, 0]);  primals_1 = None
    mm_1 = torch.ops.aten.mm.default(permute, tangents_1);  permute = None
    permute_1 = torch.ops.aten.permute.default(primals_2, [1, 0]);  primals_2 = None
    mm_2 = torch.ops.aten.mm.default(tangents_1, permute_1);  tangents_1 = permute_1 = None
    return pytree.tree_unflatten([mm, mm_2, mm_1], self._out_spec)
```
I don't have any `matmul` ops in test function, but the joint graph here has 3 `torch.ops.aten.mm`. Is it expected? Here i confused about why the AOTAutograd tracing will trace out these aten ops?
I check the second `dispatch_trace` call stack, it called from the `aot_function`.
(code here: https://github.com/pytorch/pytorch/blob/0d95378341b4eb19849295c7ccab08cc9be328a7/torch/_functorch/aot_autograd.py#L4535)

Here is the simplified backtrace for the second calling `dispatch_trace`.
  /home/pt-gpu/zejun/chille.py(11)<module>()
-> output = test(input)
..........
  /home/pt-gpu/zejun/pytorch/torch/__init__.py(1604)__call__()
-> return compile_fx(model_, inputs_, config_patches=self.config)
  /home/pt-gpu/zejun/pytorch/torch/_inductor/compile_fx.py(1196)compile_fx()
-> return aot_autograd(
  /home/pt-gpu/zejun/pytorch/torch/_dynamo/backends/common.py(56)compiler_fn()
-> cg = aot_module_simplified(gm, example_inputs, **kwargs)
  /home/pt-gpu/zejun/pytorch/torch/_functorch/aot_autograd.py(4800)aot_module_simplified()
-> compiled_fn = create_aot_dispatcher_function(
  /home/pt-gpu/zejun/pytorch/torch/_functorch/aot_autograd.py(4320)create_aot_dispatcher_function()
-> compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
  /home/pt-gpu/zejun/pytorch/torch/_functorch/aot_autograd.py(2607)aot_wrapper_dedupe()
-> return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
  /home/pt-gpu/zejun/pytorch/torch/_functorch/aot_autograd.py(2794)aot_wrapper_synthetic_base()
-> return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
  /home/pt-gpu/zejun/pytorch/torch/_functorch/aot_autograd.py(1937)aot_dispatch_base()
-> compiled_fw = compiler(fw_module, updated_flat_args)
  /home/pt-gpu/zejun/pytorch/torch/_inductor/compile_fx.py(1078)**fw_compiler_base**()
-> **joint_graph_passes(model)**
  /home/pt-gpu/zejun/pytorch/torch/_inductor/fx_passes/joint_graph.py(211)**joint_graph_passes**()
-> lazy_init()
  /home/pt-gpu/zejun/pytorch/torch/_inductor/pattern_matcher.py(1267)lazy_init()
-> result = fn()
  /home/pt-gpu/zejun/pytorch/torch/_inductor/fx_passes/joint_graph.py(32)lazy_init()
-> **_pad_mm_init**()
  /home/pt-gpu/zejun/pytorch/torch/_inductor/fx_passes/pad_mm.py(457)**_pad_mm_init**()
-> **register_replacement**(
  /home/pt-gpu/zejun/pytorch/torch/_inductor/pattern_matcher.py(923)**register_replacement**()
-> pattern = gen_pattern(
  /home/pt-gpu/anaconda3/envs/zejun/lib/python3.9/contextlib.py(79)inner()
-> return func(*args, **kwds)
  /home/pt-gpu/zejun/pytorch/torch/_inductor/pattern_matcher.py(949)gen_pattern()
-> search_gm = trace_fn(search_fn, example_inputs)
  /home/pt-gpu/zejun/pytorch/torch/utils/_contextlib.py(115)decorate_context()
-> return func(*args, **kwargs)
  /home/pt-gpu/zejun/pytorch/torch/_inductor/pattern_matcher.py(1199)training_graph()
-> aot_function(
  /home/pt-gpu/zejun/pytorch/torch/_functorch/aot_autograd.py(4627)returned_function()
-> compiled_fn = create_aot_dispatcher_function(
..........
  /home/pt-gpu/zejun/pytorch/torch/fx/experimental/proxy_tensor.py(485)**dispatch_trace**()
-> name = root.__class__.__name__ if isinstance(root, torch.nn.Module) else root.__name__


- Thank you for the explanation for the `torch._C._TorchDispatchModeKey.PROXY`, while i find in `__enter__` function of the `ProxyTorchDispatchMode`, this key is not to be set, but unset. I wonder why here call `torch._C._unset_dispatch_mode` to unset this key?
(code here: https://github.com/pytorch/pytorch/blob/0d95378341b4eb19849295c7ccab08cc9be328a7/torch/fx/experimental/proxy_tensor.py#L601)


For the issue(https://github.com/pytorch/pytorch/issues/109848), i think if there could be one more mechanism to conditionally trace the runtime ops into the joint graph when the FX interpreter runs the FX node which represents a runtime op, but the difficulty is that there is no link from the `fx.interpreter` go to the `fx.trace` for runtime ops. While for aten nodes, the `torch_dispatch` triggers the `ProxyTorchDispatchMode` to make fx.trace to trace the nodes into the graph.
(code here: https://github.com/pytorch/pytorch/blob/0d95378341b4eb19849295c7ccab08cc9be328a7/torch/fx/interpreter.py#L267)


Thank you so much :)